### PR TITLE
Bundle importer tests

### PR DESCRIPTION
**Pull request by hatched**
Add additional tests and test functionality for the bundle importer.

### CODE REVIEW

**Comment by hatched** [comment]
No QA Necessary

File: app/utils/bundle-importer.js [hunk 1]
```diff
@@ -227,6 +227,11 @@ YUI.add('bundle-importer', function(Y) {
     _executeDryRun: function(records) {
       if (this._dryRunIndex === records.length - 1) {
         // Done executing the recordSet.
+        this.db.notifications.add({
```
Comment (hatched) [inline app/utils/bundle-importer.js]: Driveby - it's nice to have a notification to tell you when it's done importing as large bundles can take a while.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2511/
Test FAILed.

File: test/data/wordpress-bundle-recordset.json [hunk 2]
```diff
@@ -0,0 +1,218 @@
+[
```
Comment (hatched) [inline test/data/wordpress-bundle-recordset.json]: recordset asset for a wordpress bundle.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2512/
Test PASSed.

**Comment by hatched** [comment]
I need this branch for my next one so I'm going to :shipit: - any comments can be done in the follow-up

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

File: test/test_bundle_importer.js [hunk 3]
```diff
@@ -19,44 +19,43 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
-        }
-      },
-      fakebackend: {
-        _loadCharm: loadCharm
-      }
+      env: env,
+      db: db,
+      fakebackend: fakebackend
+
+      // {
```
Comment (mitechie) [inline test/test_bundle_importer.js]: what is this and can it be removed?

File: test/test_bundle_importer.js [hunk 4]
```diff
@@ -19,44 +19,43 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+      //   _loadCharm: loadCharm
+      // }
     });
   });
 
   afterEach(function() {
-
+    db.destroy();
+    env.destroy();
```
Comment (mitechie) [inline test/test_bundle_importer.js]: do we need to clean up the fakebackend as well?

File: test/test_bundle_importer.js [hunk 5]
```diff
@@ -67,6 +66,15 @@ describe('Bundle Importer', function() {
   });
 
   describe('Public methods', function() {
+    describe('importBundleYAML', function() {
+      it('calls fetchDryRun with yaml', function() {
```
Comment (mitechie) [inline test/test_bundle_importer.js]: this tests with a fakebackend. Is there another test for a real back end?

File: test/test_bundle_importer.js [hunk 6]
```diff
@@ -177,13 +197,15 @@ describe('Bundle Importer', function() {
...
+      bundleImporter._execute_setAnnotations = executor;
 
       var sortedRecords = [
         { method: 'addCharm' },
         { method: 'addMachines' },
         { method: 'deploy' },
         { method: 'addUnit' },
-        { method: 'addRelation' }
+        { method: 'addRelation' },
+        { method: 'setAnnotations' }
```
Comment (mitechie) [inline test/test_bundle_importer.js]: is there a such thing as addservice or just addCharm?

File: test/test_bundle_importer.js [hunk 7]
```diff
@@ -191,26 +213,70 @@ describe('Bundle Importer', function() {
...
         deploy: true,
         addUnit: true,
-        addRelation: true
+        addRelation: true,
+        setAnnotations: true
       });
     });
-
-    it('does not fail if unknown record type', function() {
+    it('properly sorts a recordSet', function() {
```
Comment (mitechie) [inline test/test_bundle_importer.js]: can you please leave a blank line between tests? This makes it harder to find the start/end of them.

File: test/test_bundle_importer.js [hunk 8]
```diff
@@ -191,26 +213,70 @@ describe('Bundle Importer', function() {
...
+      var execute = utils.makeStubMethod(bundleImporter, '_executeRecord');
+      var notification = utils.makeStubMethod(
+          bundleImporter.db.notifications, 'add');
+      this._cleanups.concat([execute.reset, notification.reset]);
       bundleImporter._executeDryRun(sortedRecords);
+      execute.passThroughToOriginalMethod();
+      // the executor should only be called once at which time it'll throw a
+      // notification instead of continuing on.
+      assert.equal(execute.callCount(), 1, 'execute not called');
+      assert.equal(notification.callCount(), 1, 'notification not added');
```
Comment (mitechie) [inline test/test_bundle_importer.js]: is there a user facing message to check/display here?

File: test/test_bundle_importer.js [hunk 9]
```diff
@@ -191,26 +213,70 @@ describe('Bundle Importer', function() {
...
 
-    describe('_execute_addCharm', function() {
-      it('loads a charm');
-      it('shows notification on failure');
+  describe('Changeset execution', function() {
+    it('Sets up the correct environment (Integration)', function() {
+      var data = utils.loadFixture(
+          'data/wordpress-bundle-recordset.json', true);
+      bundleImporter.importBundleDryRun(data);
+      assert.equal(db.services.size(), 3);
```
Comment (mitechie) [inline test/test_bundle_importer.js]: this is then a test against the fake backend?

File: test/test_bundle_importer.js [hunk 10]
```diff
@@ -191,26 +213,70 @@ describe('Bundle Importer', function() {
...
+      assert.equal(db.machines.item(1).id, 'new1');
+      assert.equal(db.machines.item(2).id, 'new0/lxc/new3');
+      assert.equal(db.machines.item(3).id, 'new1/lxc/new2');
+      // Relations
+      assert.equal(
+          db.relations.item(0).get('id'),
+          'pending-$addService-1:reverseproxy$addService-4:website');
+      assert.equal(
+          db.relations.item(1).get('id'),
+          'pending-$addService-4:db$addService-7:db');
```
Comment (mitechie) [inline test/test_bundle_importer.js]: annotations check?

**Comment by mitechie** [comment]
a hand full of questions. Overall a :+1: with the notes addressed.

File: test/test_bundle_importer.js (hunk 10 above)
Comment (hatched) [inline test/test_bundle_importer.js]: It is known to be broken so left it out for now - will add a proper annotations test in the branch which fixes that bug.

File: test/test_bundle_importer.js (hunk 9 above)
Comment (hatched) [inline test/test_bundle_importer.js]: This test assumes that the db, env, and fakebackend are all working correctly. Each executor calls out to a number of env, db, fakebackend methods which would have made the tests very tedious as we'd have to re-implement the api and return structures.

File: test/test_bundle_importer.js (hunk 8 above)
Comment (hatched) [inline test/test_bundle_importer.js]: There is, I could add that the notification contents to the check.

File: test/test_bundle_importer.js (hunk 7 above)
Comment (hatched) [inline test/test_bundle_importer.js]: yup will do

File: test/test_bundle_importer.js (hunk 6 above)
Comment (hatched) [inline test/test_bundle_importer.js]: just addCharm add service is just called 'deploy'. In the future we will also want to add addCharm functionality to the environment instead of relying on 'deploy'. Right now we call the fakebackend directly to support it in the sandbox mode.

File: test/test_bundle_importer.js (hunk 3 above)
Comment (hatched) [inline test/test_bundle_importer.js]: Oops yes this will be removed

File: test/test_bundle_importer.js (hunk 4 above)
Comment (hatched) [inline test/test_bundle_importer.js]: can't hurt, I'll add that.

File: test/test_bundle_importer.js (hunk 5 above)
Comment (hatched) [inline test/test_bundle_importer.js]: This test actually doesn't do a whole lot. It's just checking that calling the public `importBundleYAML` method calls the `fetchDryRun` method. The `fetchDryRun` method tests are further down in the suite.

File: test/test_bundle_importer.js (hunk 10 above)
Comment (mitechie) [inline test/test_bundle_importer.js]: Can you do an XXX here then with the known issue and then remove it in your follow up branch?
